src/webauthn.rs

```diff
@@ -0,0 +1,88 @@
+use anyhow::Context;
+use reqwest::Url;
+use webauthn_authenticator_rs::{
+    ctap2::EnrollSampleStatus,
+    transport::{AnyTransport, Transport},
+    types::{CableRequestType, CableState},
+    ui::UiCallback,
+    AuthenticatorBackend,
+};
+use webauthn_rs_proto::PublicKeyCredentialRequestOptions;
+
+use crate::locked::Password;
+
+pub async fn webauthn(
+    challenge: PublicKeyCredentialRequestOptions,
+    pin: &str,
+) -> Result<Password, Box<dyn std::error::Error + Send + Sync>> {
+    let mut trans = AnyTransport::new()
+        .await
+        .ok()
+        .ok_or("Failed to set up webauthn transport")?;
+
+    let ui = Pinentry {
+        pin: pin.to_string(),
+    };
+    let mut authenticator = trans
+        .connect_all(&ui)
+        .ok()
+        .ok_or("Failed to connect to authenticator")?;
+    let authenticator = authenticator
+        .get_mut(0)
+        .ok_or("Failed to get authenticator")?;
+
+    let origin = crate::config::Config::load_async()
+        .await
+        .ok()
+        .ok_or("error loading config")?
+        .base_url
+        .ok_or("error loading base_url")?;
+    let origin = Url::parse(&origin).context("Failed to parse origin")?;
+
+    let result = authenticator.perform_auth(origin, challenge, 60000);
+    // required, so that the JSON is parsed corretly by the server
+    let out = serde_json::to_string(&result.unwrap())?
+        .replace("\"appid\":null,\"hmac_get_secret\":null", "\"appid\":false")
+        .replace("clientDataJSON", "clientDataJson");
```

Comment:
This looks quite brittle.

An alternative approach would be to create a wrapper struct which serializes the fields in the way the API expects them.

Reply:
Yeah, agreed, your proposed solution makes sense. I'll change it to that.